RecoBTag/FeatureTools/plugins/ParticleTransformerAK4TagInfoProducer.cc

```diff
@@ -0,0 +1,427 @@
+  typedef reco::VertexCollection VertexCollection;
+
+  void beginStream(edm::StreamID) override {}
+  void produce(edm::Event&, const edm::EventSetup&) override;
+  void endStream() override {}
+
+  const double jet_radius_;
+  const double min_candidate_pt_;
+  const bool flip_;
+
+  edm::EDGetTokenT<edm::View<reco::Jet>> jet_token_;
+  edm::EDGetTokenT<VertexCollection> vtx_token_;
+  edm::EDGetTokenT<SVCollection> sv_token_;
+  edm::EDGetTokenT<edm::ValueMap<float>> puppi_value_map_token_;
+  edm::EDGetTokenT<edm::ValueMap<int>> pvasq_value_map_token_;
+  edm::EDGetTokenT<edm::Association<VertexCollection>> pvas_token_;
+  edm::EDGetTokenT<edm::View<reco::Candidate>> candidateToken_;
+  edm::ESGetToken<TransientTrackBuilder, TransientTrackRecord> track_builder_token_;
+  bool use_puppi_value_map_;
+  bool use_pvasq_value_map_;
+
+  bool fallback_puppi_weight_;
+  bool fallback_vertex_association_;
+
+  bool is_weighted_jet_;
+
+  const double min_jet_pt_;
+  const double max_jet_eta_;
+};
+
+ParticleTransformerAK4TagInfoProducer::ParticleTransformerAK4TagInfoProducer(const edm::ParameterSet& iConfig)
+    : jet_radius_(iConfig.getParameter<double>("jet_radius")),
+      min_candidate_pt_(iConfig.getParameter<double>("min_candidate_pt")),
+      flip_(iConfig.getParameter<bool>("flip")),
+      jet_token_(consumes<edm::View<reco::Jet>>(iConfig.getParameter<edm::InputTag>("jets"))),
+      vtx_token_(consumes<VertexCollection>(iConfig.getParameter<edm::InputTag>("vertices"))),
+      sv_token_(consumes<SVCollection>(iConfig.getParameter<edm::InputTag>("secondary_vertices"))),
+      candidateToken_(consumes<edm::View<reco::Candidate>>(iConfig.getParameter<edm::InputTag>("candidates"))),
+      track_builder_token_(
+          esConsumes<TransientTrackBuilder, TransientTrackRecord>(edm::ESInputTag("", "TransientTrackBuilder"))),
+      use_puppi_value_map_(false),
+      use_pvasq_value_map_(false),
+      fallback_puppi_weight_(iConfig.getParameter<bool>("fallback_puppi_weight")),
+      fallback_vertex_association_(iConfig.getParameter<bool>("fallback_vertex_association")),
+      is_weighted_jet_(iConfig.getParameter<bool>("is_weighted_jet")),
+      min_jet_pt_(iConfig.getParameter<double>("min_jet_pt")),
+      max_jet_eta_(iConfig.getParameter<double>("max_jet_eta")) {
+  produces<ParticleTransformerAK4TagInfoCollection>();
+
+  const auto& puppi_value_map_tag = iConfig.getParameter<edm::InputTag>("puppi_value_map");
+  if (!puppi_value_map_tag.label().empty()) {
+    puppi_value_map_token_ = consumes<edm::ValueMap<float>>(puppi_value_map_tag);
+    use_puppi_value_map_ = true;
+  } else if (is_weighted_jet_) {
+    throw edm::Exception(edm::errors::Configuration,
+                         "puppi_value_map is not set but jet is weighted. Must set puppi_value_map.");
+  }
+
+  const auto& pvas_tag = iConfig.getParameter<edm::InputTag>("vertex_associator");
+  if (!pvas_tag.label().empty()) {
```

Comment:
`vertex_associator` is defined in the fillDescription: how can it be empty?
Unless I miss something, please move all definitions into the class initialization